File: application/service/Decorate/Params/Equipment/Armor.py
from models.Equipment.Armor import Common

import helpers.mongo

from models.User.Factory import User_Factory


class Decorate(object):
    def _fixParams(self, data):
        assert(data['type'] in [Common.TYPE_ARMOR_LEATHER, Common.TYPE_ARMOR_MAIL, Common.TYPE_ARMOR_PLATE])

        if data['shield']:
            assert(data['shield_type'] in [Common.TYPE_SHIELD_WOOD, Common.TYPE_SHIELD_STEEL])

        if 'user' in data:
            data['user'] = helpers.mongo.objectId(data['user'])

        data['health'] = int(data['health'])
        data['agility'] = int(data['agility'])
        data['absorption'] = int(data['absorption'])

        data['shield'] = bool(data['shield'])
        data['shield_type'] = data['shield_type']
        data['shield_blocking'] = int(data['shield_blocking'])
        data['shield_durability'] = int(data['shield_durability'])

        return data

    def save(self, data, user=None):
        data = self._fixParams(data)
        return super().save(data, user)

    def simulate(self, data):
        data = self._fixParams(data)
        return super().simulate(data)

    def get(self, _id, user=None):
        _id = helpers.mongo.objectId(_id)
        return super().get(_id, user)

    def load(self, user):
        userId = helpers.mongo.objectId(user)
        user = User_Factory.getDomainById(userId)
        return super().load(user)

    def remove(self, _id, user=None):
        return super().remove(
            helpers.mongo.objectId(_id),
            user
        )

File: public/js/model/equipment/armor.js
define('model/equipment/armor', [
], function () {
    return AbstractModel.extend({
        model_url: "equipment/armor",

        load: function (successCallback) {
            this.sync('get', {
                data: {
                    _id: this.get('_id')
                },
                success: successCallback
            });
        },

        save: function (successCallback) {
            this.sync('save', {
                data: {
                    health: this.get('health'),
                    agility: this.get('agility'),
                    absorption: this.get('absorption'),

                    shield: this.get('shield'),
                    shield_type: this.get('shield_type'),
                    shield_durability: this.get('shield_durability'),
                    shield_blocking: this.get('shield_blocking'),

                    type: this.get('type'),
                    user: this.get('user')
                },
                success: successCallback
            });
        },

        simulate: function () {
            if (
                (!this.get('health') && !_.isNumber(this.get('health'))) ||
                (!this.get('agility') && !_.isNumber(this.get('agility'))) ||
                (!this.get('absorption') && !_.isNumber(this.get('absorption')))
            ) {
                return;
            } else {
                if (
                    this.get('shield') &&
                    (
                        (!this.get('shield_type') && _.isNumber(this.get('shield_type'))) ||
                        (!this.get('shield_durability') && _.isNumber(this.get('shield_durability'))) ||
                        (!this.get('shield_blocking') && _.isNumber(this.get('shield_blocking')))
                    )
                ) {
                    return;
                }
            }

            this.sync('simulate', {
                data: this.attributes
            });
        },

        remove: function (success) {
            this.sync('remove', {
                data: {
                    '_id': this.get('_id')
                },
                success: success
            });
        }
    });
});